Register values listed in __constants__ as attributes of the Module.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26584 [quant] Un-hardcode epsilon constant in FoldConvBatchNorm2d.
* **#26581 Register values listed in __constants__ as attributes of the Module.**

We're currently inlining immediate values of the constants directly into
IR when we generate it providing no way to access these values by their
names later. This change registers such values as atrtibutes of the
module so that they are not lost after IR generation.

Differential Revision: [D17513451](https://our.internmc.facebook.com/intern/diff/D17513451)

Commit:
Register values listed in __constant__ as attributes of the Module.

We're currently inlining immediate values of the constants directly into
IR when we generate it providing no way to access these values by their
names later. This change registers such values as atrtibutes of the
module so that they are not lost after IR generation.

[ghstack-poisoned]

Commit:
Love new inline linter! on "Register values listed in __constants__ as attributes of the Module."


We're currently inlining immediate values of the constants directly into
IR when we generate it providing no way to access these values by their
names later. This change registers such values as atrtibutes of the
module so that they are not lost after IR generation.

Differential Revision: [D17513451](https://our.internmc.facebook.com/intern/diff/D17513451)

[ghstack-poisoned]

suo reviewed (changes requested):
I am about to put up a substantial rewrite of ModuleValue and associated logic, so I think it would be better  to base these changes off of that. In particular, we are (for the most part) going to disallow `py_module_` lookups in `ModuleValue::attr`

jerryzh168 commented:
@suo is that going to be done before branch cut(9/24)?

suo commented:
No. Is this functionality needed for the release?

jerryzh168 commented:
> No. Is this functionality needed for the release?

I think so, otherwise BatchNorm folding is not going to work properly

Commit:
Add tests to verify that we can't mutate constant attributes. on "Register values listed in __constants__ as attributes of the Module."


We're currently inlining immediate values of the constants directly into
IR when we generate it providing no way to access these values by their
names later. This change registers such values as atrtibutes of the
module so that they are not lost after IR generation.

Differential Revision: [D17513451](https://our.internmc.facebook.com/intern/diff/D17513451)

[ghstack-poisoned]

suo commented on torch/csrc/jit/script/python_sugared_value.cpp:
```diff
@@ -447,6 +449,24 @@ std::shared_ptr<SugaredValue> ModuleValue::attr(
+    // Values of the attributes listed in the _constants_set will be put
+    // directly into IR. In order to allow us to access these values by their
+    // name after IR is generated, we register them as attributes.
+    if (py::isinstance<py::bool_>(attr)) {
```
`pybind_utils.h` has the functions `tryToInferType` and `toIValue`, which can be combined to make this more general.

Commit:
Update on "Register values listed in __constants__ as attributes of the Module."


We're currently inlining immediate values of the constants directly into
IR when we generate it providing no way to access these values by their
names later. This change registers such values as atrtibutes of the
module so that they are not lost after IR generation.

Differential Revision: [D17513451](https://our.internmc.facebook.com/intern/diff/D17513451)

[ghstack-poisoned]

suo commented:
this PR breaks mkldnn tests on master
